### PR TITLE
Rubyfied version of the ES SSM patch script

### PR DESCRIPTION
**Pull request by andrejbl**
This PR contains a rubyfied version of https://github.com/intercom/elasticsearch6-cookbook/blob/master/baseline_patch.sh with added unit tests and some boilerplate configs.

### CODE REVIEW

**Comment by ccoffey** [comment]
Can you please add `Responsible Team: Team Datastores` as the first line of the README?

Similar to what we have for Magpie: https://github.com/intercom/magpie

File: lib/es_client.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    private
+
+    def update_es_settings(settings)
+      response = Typhoeus::Request.new(es_settings_url, method: :put, headers: { 'Content-Type': 'application/json' }, body: settings).run.response_body
+      Logger.log(response.to_s)
+
+      return if JSON.parse(response)['acknowledged'] == true
+
+      raise PatchError.new("Error updating settings to #{settings}: #{response}", 1)
```
[the remark this answers is not in the document]
Comment (andrejbl) [reply lib/es_client.rb]: If you look closely, the response is logged on line 53 above in all cases.

File: lib/logger.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+require 'date'
+
+module ElasticsearchSSM
+  #
+  # Logs ES SSM command runs to stdout
+  # Stdout is published to S3 by SSM
```
[the remark this answers is not in the document]
Comment (andrejbl) [reply lib/logger.rb]: The original SSM bash script included logging to a local file. However, in practice, I've never had a need to use it - SSM logs all stdout to an S3 file and provides an easy link to follow to it from within the SSM UI (and the file is text therefore viewable in the browser directly). I will keep this as is, we can add the file logging or Cloudwatch logging if we ever see a need for it.

File: lib/patch_client.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      begin
+        es_client.wait_for_green_state
+        exit_code = safe_reboot if az_schedule_match? && run_security_update
+        return exit_code
+      rescue PatchError => e
+        Logger.log("#{e.message} - #{e.error_details}")
+        exit_code = 1
+      ensure
+        es_client.enable_shard_allocation if exit_code != AWS_SSM_REBOOT_CODE
+      end
```
[the remark this answers is not in the document]
Comment (andrejbl) [reply lib/patch_client.rb]: Having the logic in its separate class makes it better encapsulated and much easier to unit test, as you don't need to worry about pesky Kernel.exit() calls which I am loathe to mock in unit tests. It also allows the bin script to be a simple dumb wrapper that doesn't need tests.

File: lib/es_client.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+require 'typhoeus'
+require 'json'
+
+module ElasticsearchSSM
+  #
+  # Simple ES transport client based on Typhoeus
```
[the remark this answers is not in the document]
Comment (andrejbl) [reply lib/es_client.rb]: When I first started writing this code, I had the same idea, but found the elasticsearch-ruby gem documentation to be awful (including RubyDoc) so I just abandoned the approach alltogether as this is a relatively simple use case. Trying a second time, I actually downloaded the code and did a bit of reading, and the documentation inside is a bit easier to follow, so I will try to refactor this to use the gem. I will keep the ESClient layer as a wrapper to still to keep patching logic code simple.

File: lib/es_client.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    def initialize(es_endpoint = 'http://localhost:9200')
+      @elasticsearch_client = Elasticsearch::Client.new(url: es_endpoint, log: true)
+    end
+
+    def cluster_green?
+      elasticsearch_client.cluster.health["status"] == 'green'
+    end
+
+    def enable_shard_allocation
+      response = elasticsearch_client.cluster.put_settings(body: { transient: { 'cluster.routing.allocation.enable' => nil } })
```
Comment (ccoffey) [inline lib/es_client.rb]: Setting this property to nil in order to enable it is a little weird. Can we instead set it to `all`?
Ref: https://www.elastic.co/guide/en/elasticsearch/reference/current/shards-allocation.html

File: lib/es_client.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  # Slim wrapper for common ES operations
+  #
+  class ESClient
+    attr_reader :elasticsearch_client
+
+    def initialize(es_endpoint = 'http://localhost:9200')
+      @elasticsearch_client = Elasticsearch::Client.new(url: es_endpoint, log: true)
+    end
+
+    def cluster_green?
```
Comment (ccoffey) [inline lib/es_client.rb]: I like that this method ends in `?` to indicate that it returns a boolean. 

How about changing `enable_shard_allocation` to `enable_shard_allocation!` to indicate that it modifies external state? Same comment for `disable_shard_allocation!` and `stop_elasticsearch!`

File: lib/es_client.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      raise PatchError.new("Error enabling shard allocation: #{response}", 1) if response["acknowledged"] != true
+    end
+
+    def disable_shard_allocation
+      response = elasticsearch_client.cluster.put_settings(body: { transient: { 'cluster.routing.allocation.enable' => 'none' } })
+      raise PatchError.new("Error disabling shard allocation: #{response}", 1) if response["acknowledged"] != true
+    end
+
+    def stop_elasticsearch
+      ShellClient.run_shell_command('sudo monit stop elasticsearch')
```
Comment (ccoffey) [inline lib/es_client.rb]: Is there anyway to indicate if this succeeded or failed? What would happen if this failed silently? Would we end up installing patches but not restarting the host? Is this ok?

File: lib/patch_client.rb [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    def initialize(es_client)
+      @es_client = es_client
+    end
+
+    def run
+      exit_code = 0
+      begin
+        es_client.wait_for_green_state
+        exit_code = safe_reboot if az_schedule_match? && run_security_update
```
[the remark this answers is not in the document]
Comment (andrejbl) [reply lib/patch_client.rb]: That is actually a good point and I will refactor. I am sometimes to hooked up on getting a minimal amount of ruby code :)

File: lib/es_client.rb (hunk 7 above)
Comment (andrejbl) [reply lib/es_client.rb]: Monit itself does not return any output on success (if you ran this command in shell yourself you wouldn't get anything. I will add -v, the stdout is logged.

For failures, ShellClient raises a `PatchError` and corresponding outputs and status codes are logged. So this cannot fail silently.

File: lib/es_client.rb (hunk 5 above)
Comment (andrejbl) [reply lib/es_client.rb]: Nil should set it to whatever is the default, so it is a safer option (let's say we disabled shard allocation for a cluster for investigation and still need to install a critical update).

File: lib/es_client.rb (hunk 6 above)
Comment (andrejbl) [reply lib/es_client.rb]: I wouldn't go that far. ES client doesn't call put_settings `put_settings!` even though it changes the settings. ! is more for methods that destroy or alter object state in memory, not remote one imo.

**Comment by andrejbl** [comment]
> I assume that is a wrapper that checks this script's exit code and performs reboot if it's value is `AWS_SSM_REBOOT_CODE`, is this right? Why don't we initiate reboot directly from the script?

Reboot is initiated by the SSM agent running on the machine after it receives the exit code, so the code itself doesn't have to do anything.

**Pull request merged**